src/legacy/core_plugins/interpreter/types/types.ts

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * A generic type which represents a custom Expression Type Definition that's
+ * registered to the Interpreter.
+ */
+export interface ExpressionTypeDef<Name extends string, Type, SerializedType = undefined> {
```

Comment:
nit: this is personal style, but I don't care for this name.  Seeing it in other files had me thinking it meant type as in Typescript and def was redundant.  Naming is the hardest thing, so it's up to you.

Reply:
Ah this is a tough one; I thought the same thing about potential for confusion when using "def". I had changed it because @w33ble pointed out that elsewhere in Canvas, we refer to these as the Type "Definitions"; I'm happy to revert though, as I don't feel strongly one way or the other.

Reply:
I'm torn, honestly.  Some other options that might be less confusing:

- Expression (isn't that really what this is?)
- ExpressionDT
- ExpressionTypeDefinition

Reply:
> Expression (isn't that really what this is?)

No, this is a definition for the various data types that the expression handles. And there's a distinction between a type's definition (the object this interface is defining) and it's instance (what is provided by the TypeRegistry). 

I mentioned preferring adding something extra to the `ExpressionType` name, since there will also be an instance of ExpressionType that is created using this definition.